DQMOffline/Trigger/python/ZGammaplusJetsMonitor_cff.py

```diff
@@ -0,0 +1,67 @@
+import FWCore.ParameterSet.Config as cms
+
+from DQMOffline.Trigger.ZGammaplusJetsMonitor_cfi import hltZJetsmonitoring
+from JetMETCorrections.Configuration.JetCorrectors_cff import *
+
+ak4PFJetsPuppiCorrected = cms.EDProducer('CorrectedPFJetProducer',
+    src                 = cms.InputTag('ak4PFJetsPuppi'),
```

Comment:
this product seems to be not always available, see e.g. failures in wf [136.7801](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a9e10a/38155/runTheMatrixINPUT-results/136.7801_RunHLTPhy2017B_AOD/step2_RunHLTPhy2017B_AOD.log) and [136.7803](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a9e10a/38155/runTheMatrixINPUT-results/136.7803_RunHLTPhy2017B_RAWAOD/step2_RunHLTPhy2017B_RAWAOD.log)

Reply:
Hi! Thanks for the comments! I will change them. Yes, it targets 2024 pp data-taking, sorry I was not sure for the backport.

Reply:
Hi! I want to ask since this happened because Puppi Jets were not available in the input sample, but if the producer were not in cff it would not give an error. This is something we should avoid in general when using corrections in DQM? What I did is use corrections on the fly. I rerun and took same distributions and also do the code checks. 
Thanks, Eirini.

Reply:
@esiam 

>  I rerun and took same distributions and also do the code checks.

thank you for the updates.

>  I want to ask since this happened because Puppi Jets were not available in the input sample, but if the producer were not in cff it would not give an error. This is something we should avoid in general when using corrections in DQM? 

I know very little of the underlying motivation behind this PR, so I don't feel qualified to comment if you should "in general" avoid these corrections.
Allow me to point out the following facts:
   * all the other test workflows actually succeeded: https://github.com/cms-sw/cmssw/pull/44411#issuecomment-2000011303
   * the only 2 workflow with problems  [136.7801](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a9e10a/38155/runTheMatrixINPUT-results/136.7801_RunHLTPhy2017B_AOD/step2_RunHLTPhy2017B_AOD.log) and [136.7803](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a9e10a/38155/runTheMatrixINPUT-results/136.7803_RunHLTPhy2017B_RAWAOD/step2_RunHLTPhy2017B_RAWAOD.log) are starting from AOD and not from RAW data [1]

In both workflows the DQM sequence run is  `DQM:offlineHLTSourceOnAOD` which is defined here https://github.com/cms-sw/cmssw/blob/8731f56052dcd23f22252b9d90be28cf4cd2f044/DQMOffline/Trigger/python/DQMOffline_Trigger_cff.py#L124-L146

my naive intuition would be that **in case you want to run the corrector for the worfklows starting from RAW**, you can modify the definition of  `offlineHLTSourceOnAOD` such that instead of running the sequence `jetmetMonitorHLT` runs a different sequence (same as the regular one, but without the crashing producer, or a modified producer that could work on AOD).
Please let me know if that makes sense to you.
  
   
[1]
```
$ runTheMatrix.py -nel 136.7801,136.7803

136.7801 RunHLTPhy2017B_AOD+DQMHLTonAOD_2017+HARVESTDQMHLTonAOD_2017 [1]: input from: /HLTPhysics/Run2017B-PromptReco-v1/AOD with run [] 
                                           [2]: cmsDriver.py step2  -s DQM:offlineHLTSourceOnAOD --conditions auto:run2_data --data  --eventcontent DQM --datatier DQMIO --era Run2_2017 --fileout DQMHLTonAOD.root --procModifiers tau_readOldDiscriminatorFormat -n 100 
                                           [3]: cmsDriver.py step3  -s HARVESTING:hltOfflineDQMClient --conditions auto:run2_data --data  --filetype DQM --scenario pp --era Run2_2017 --filein file:DQMHLTonAOD.root -n 100 

136.7803 RunHLTPhy2017B_RAWAOD+DQMHLTonRAWAOD_2017+HARVESTDQMHLTonAOD_2017 [1]: input from: /HLTPhysics/Run2017B-PromptReco-v1/AOD with run [] 
                                           [2]: cmsDriver.py step2  --process reHLT -s HLT:@relval2017,DQM:offlineHLTSourceOnAOD --conditions auto:run2_data --data  --eventcontent DQM --datatier DQMIO --era Run2_2017 --secondfilein filelist:step1_dasparentquery.log --customise_commands "process.HLTAnalyzerEndpath.remove(process.hltL1TGlobalSummary)" --fileout DQMHLTonAOD.root --procModifiers tau_readOldDiscriminatorFormat -n 100 
                                           [3]: cmsDriver.py step3  -s HARVESTING:hltOfflineDQMClient --conditions auto:run2_data --data  --filetype DQM --scenario pp --era Run2_2017 --filein file:DQMHLTonAOD.root -n 100 

2 workflows with 3 steps
```

Reply:
Hi! and thanks for the reply!! 
To me (I mean how I understand this fail) the problem is that the ak4PFJetsPuppi collection is not there. The exception in the log file is the following: 

----- Begin Fatal Exception 15-Mar-2024 11:53:55 CET-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 297227 lumi: 1 event: 53500 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Calling method for module CorrectedPFJetProducer/'ak4PFJetsPuppiCorrected'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::PFJet>
Looking for module label: ak4PFJetsPuppi
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------

Because of the producer. But if for example I use another (not puppi) Jet collection this would not complain. So in order to be sure that in case that our collection is missing, it would not give a fatal exception we should not use the producer and use another way to apply corrections or modify it. But we (JME) want to use Puppi for offline monitoring so, I want to ask because I do not know are they not included in AOD files or this fail is because this is an older workflow and they (Puppi Jets) were not included then? 
Thanks, Eirini.

Reply:
>  I want to ask because I do not know are they not included in AOD files or this fail is because this is an older workflow and they (Puppi Jets) were not included then?

I think this question should be answered by @cms-sw/jetmet-pog-l2 . I see that the failing workflows are from Run2 (2017). If back then puppi jets were not produced it would be understood why the collection is not available in old AOD data . Let me ask again if you really want this to run this on workflows starting from AOD. If yes, I guess you could use an era modifier to remove the producer in the case of 2017 (or other years), otherwise the suggestion at https://github.com/cms-sw/cmssw/pull/44411#discussion_r1527154927 still holds.

Reply:
Hi and thanks for your reply. In my opinion, I think we should keep the sequence there. By applying the offline corrections to be calculated on the fly (after the latest push)
- runTheMatrix.py -l limited -i all --ibeos and the two that had the fatal exception
- runTheMatrix.py -l 136.7803 --ibeos
- runTheMatrix.py -l 136.7801 --ibeos  are passing. So, for now I think it is ok. 

For the second question I asked, I was trying to understand why the offline Puppi collections were missing and if this is because of the file type AOD.

Thanks a lot for your commends and time,
Eirini.

Reply:
>  In my opinion, I think we should keep the sequence there.

removing the sequence was never an option. The suggestion was to modify it such that it can run also on old 2017 AOD data.

Reply:
Hello @mmusich , thanks for all the comments. 
I think there is a difference in what each calls sequence above, and just to make sure we have converged on this point I am summarizing the situation : 

What Rena was referring above by "removing the sequence" to removing the "corrections sequence"  of [ak4PFPuppiL1FastL2L3CorrectorChain](https://github.com/esiam/cmssw/blob/93924d04f72f24536ed02a6ebcfa6b4f8e5b7e75/JetMETCorrections/Configuration/python/JetCorrectors_cff.py#L350) from [here](https://github.com/esiam/cmssw/blob/JME_ZGammaplusJetMonitor/DQMOffline/Trigger/python/ZGammaplusJetsMonitor_cff.py#L59-L64).
So in this sense after her latest push this is not needed to run. The reason is because the "correctors" part will always work even if the `ak4PFJetsPuppi` does not exist. It only needs the `AK4PFPuppi` tag in the GT (but that should anyway exist in any GT running in 14_X otherwise all correctors that use it in CMSSW would crash).

Now then in the latest push that she did, the need to have `ak4PFJetsPuppi` is treated by not producing the corrected jets though the "CorrectedJetsProducer" which would need the collection to work (so in practice modified this sequence for DQM to work in 2017 - Marco's suggestion). The correction is applied then inside the `.cc` part of the DQM code. If the collection does not exist in this code, like in 2017 AOD files, DQM will just return with no crash.